sdk/cosmos/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java

```diff
@@ -754,91 +754,65 @@ private <T> Slice<T> sliceQuery(SqlQuerySpec querySpec,
-                .byPage(feedResponseContentSize);
-        }
-
-        final FeedResponse<JsonNode> feedResponse = feedResponseFlux
-            .publishOn(Schedulers.parallel())
-            .doOnNext(propertiesFeedResponse ->
-                CosmosUtils.fillAndProcessResponseDiagnostics(this.responseDiagnosticsProcessor,
-                    propertiesFeedResponse.getCosmosDiagnostics(), propertiesFeedResponse))
-            .onErrorResume(throwable ->
-                CosmosExceptionUtils.exceptionHandler("Failed to query items", throwable,
-                    this.responseDiagnosticsProcessor))
-            .next()
-            .block();
-
-        assert feedResponse != null;
-        final Iterator<JsonNode> it = feedResponse.getResults().iterator();
-
-        /*
+         *
          * We only use offset on the first page because of the use of continuation tokens.
          * After we apply the offset to the first page, the continuation token will pick
          * up the second and future pages at the correct index.
          */
+        int feedResponseContentSize = pageable.getPageSize();
+        String continuationToken = null;
         int offsetForPageWithoutContToken = 0;
         if (pageable instanceof CosmosPageRequest) {
             if (((CosmosPageRequest) pageable).getRequestContinuation() == null) {
+                feedResponseContentSize = (int) (feedResponseContentSize
+                    + (feedResponseContentSize * pageable.getPageNumber()) + pageable.getOffset());
                 offsetForPageWithoutContToken = (pageable.getPageNumber() * pageable.getPageSize())
                     + (int) pageable.getOffset();
             }
+            continuationToken = ((CosmosPageRequest) pageable).getRequestContinuation();
         } else {
+            feedResponseContentSize = feedResponseContentSize + (feedResponseContentSize * pageable.getPageNumber());
             offsetForPageWithoutContToken = (pageable.getPageNumber() * pageable.getPageSize());
         }
 
         final List<T> result = new ArrayList<>();
-        for (int index = 0; it.hasNext()
-            && index < pageable.getPageSize() + offsetForPageWithoutContToken; index++) {
-
-            final JsonNode jsonNode = it.next();
-            if (jsonNode == null) {
-                continue;
-            }
-
-            if (index >= offsetForPageWithoutContToken) {
-                maybeEmitEvent(new AfterLoadEvent<>(jsonNode, returnType, containerName));
-                final T entity = mappingCosmosConverter.read(returnType, jsonNode);
-                result.add(entity);
-            }
-        }
+        do {
+            feedResponseFlux = container
+                .queryItems(querySpec, cosmosQueryRequestOptions, JsonNode.class)
+                .byPage(continuationToken, feedResponseContentSize);
+            FeedResponse<JsonNode> feedResponse = generateFeedResponse(feedResponseFlux);
```

Comment:
Since we are only calling this once now, I would recommend moving it back to an inline function instead of extra function call, and clubbing it with byPage call above.